Definition/Untyped.agda

```diff
@@ -74,121 +169,182 @@ data Kind : (ns : List Nat) → Set a where
+  natrec p q r (toTerm A) (toTerm z) (toTerm s) (toTerm n)
+toTerm (gen (Unitkind s l) []) =
+  Unit s l
+toTerm (gen (Starkind s l) []) =
+  star s l
+toTerm (gen (Unitreckind l p q) (A ∷ₜ t ∷ₜ u ∷ₜ [])) =
+  unitrec l p q (toTerm A) (toTerm t) (toTerm u)
+toTerm (gen Emptykind []) =
+  Empty
+toTerm (gen (Emptyreckind p) (A ∷ₜ t ∷ₜ [])) =
+  emptyrec p (toTerm A) (toTerm t)
+toTerm (gen Idkind (A ∷ₜ t ∷ₜ u ∷ₜ [])) =
+  Id (toTerm A) (toTerm t) (toTerm u)
+toTerm (gen Reflkind []) =
+  rfl
+toTerm (gen (Jkind p q) (A ∷ₜ t ∷ₜ B ∷ₜ u ∷ₜ v ∷ₜ w ∷ₜ [])) =
+  J p q (toTerm A) (toTerm t) (toTerm B) (toTerm u) (toTerm v) (toTerm w)
+toTerm (gen (Kkind p) (A ∷ₜ t ∷ₜ B ∷ₜ u ∷ₜ v ∷ₜ [])) =
+  K p (toTerm A) (toTerm t) (toTerm B) (toTerm u) (toTerm v)
+toTerm (gen (Boxcongkind s) (A ∷ₜ t ∷ₜ u ∷ₜ v ∷ₜ [])) =
+  []-cong s (toTerm A) (toTerm t) (toTerm u) (toTerm v)
+
+-- Converting to the alternative syntax.
+
+fromTerm : Term n → Term′ n
+fromTerm (var x) =
+  var x
+fromTerm (U l) =
+  gen (Ukind l) []
+fromTerm (ΠΣ⟨ b ⟩ p , q ▷ A ▹ B) =
+  gen (Binderkind b p q) (fromTerm A ∷ₜ fromTerm B ∷ₜ [])
+fromTerm (lam p t) =
+  gen (Lamkind p) (fromTerm t ∷ₜ [])
+fromTerm (t ∘⟨ p ⟩ u) =
+  gen (Appkind p) (fromTerm t ∷ₜ fromTerm u ∷ₜ [])
+fromTerm (prod s p t u) =
+  gen (Prodkind s p) (fromTerm t ∷ₜ fromTerm u ∷ₜ [])
+fromTerm (fst p t) =
+  gen (Fstkind p) (fromTerm t ∷ₜ [])
+fromTerm (snd p t) =
+  gen (Sndkind p) (fromTerm t ∷ₜ [])
+fromTerm (prodrec r p q A t u) =
+  gen (Prodreckind r p q)
+    (fromTerm A ∷ₜ fromTerm t ∷ₜ fromTerm u ∷ₜ [])
+fromTerm ℕ =
+  gen Natkind []
+fromTerm zero =
+  gen Zerokind []
+fromTerm (suc t) =
+  gen Suckind (fromTerm t ∷ₜ [])
+fromTerm (natrec p q r A z s n) =
+  gen (Natreckind p q r)
+    (fromTerm A ∷ₜ fromTerm z ∷ₜ fromTerm s ∷ₜ fromTerm n ∷ₜ [])
+fromTerm (Unit s l) =
+  gen (Unitkind s l) []
+fromTerm (star s l) =
+  gen (Starkind s l) []
+fromTerm (unitrec l p q A t u) =
+  gen (Unitreckind l p q)
+    (fromTerm A ∷ₜ fromTerm t ∷ₜ fromTerm u ∷ₜ [])
+fromTerm Empty =
+  gen Emptykind []
+fromTerm (emptyrec p A t) =
+  gen (Emptyreckind p) (fromTerm A ∷ₜ fromTerm t ∷ₜ [])
+fromTerm (Id A t u) =
+  gen Idkind (fromTerm A ∷ₜ fromTerm t ∷ₜ fromTerm u ∷ₜ [])
+fromTerm rfl =
+  gen Reflkind []
+fromTerm (J p q A t B u v w) =
+  gen (Jkind p q)
+    (fromTerm A ∷ₜ fromTerm t ∷ₜ fromTerm B ∷ₜ fromTerm u
+                ∷ₜ fromTerm v ∷ₜ fromTerm w ∷ₜ [])
+fromTerm (K p A t B u v) =
+  gen (Kkind p)
+    (fromTerm A ∷ₜ fromTerm t ∷ₜ fromTerm B
+                ∷ₜ fromTerm u ∷ₜ fromTerm v ∷ₜ [])
+fromTerm ([]-cong s A t u v) =
+  gen (Boxcongkind s)
+    (fromTerm A ∷ₜ fromTerm t ∷ₜ fromTerm u
+                ∷ₜ fromTerm v ∷ₜ [])
 
 ------------------------------------------------------------------------
 -- Weakening
 
-  -- Weakening of terms.
-  -- If η : Γ ≤ Δ and Δ ⊢ t : A then Γ ⊢ wk η t : wk η A.
+-- Weakening of terms.
+-- If η : Γ ≤ Δ and Δ ⊢ t : A then Γ ⊢ wk η t : wk η A.
```

Comment:
Why is this not defined using `wk′`, `toTerm` and `fromTerm`?

Reply:
Originally, I defined both weakening and substitution like that but this caused `(var x) [ σ ] ≡ σ x` to no longer hold definitionally so I decided against it. But I suppose it should still be fine to define weakenings that way.

Reply:
Trying this seems to make typechecking a little slower so I'm going to leave it like this